common/src/main/java/org/opensearch/ml/common/transport/memorycontainer/memory/MLCreateEventInput.java

```diff
@@ -33,76 +28,96 @@
+    private Map<String, Object> data;
 
     // Optional fields
     private Map<String, String> namespace;
     private boolean infer;
     private Map<String, String> metadata;
-    private Map<String, String> tags;
 
-    public MLAddMemoriesInput(
+    public MLCreateEventInput(
         String memoryContainerId,
         ShortTermMemoryType memoryType,
         List<MessageInput> messages,
-        String binaryData,
-        Map<String, Object> structuredData,
+        Map<String, Object> data,
         Map<String, String> namespace,
         boolean infer,
-        Map<String, String> metadata,
-        Map<String, String> tags
+        Map<String, String> metadata
     ) {
         // MAX_MESSAGES_PER_REQUEST limit removed for performance testing
 
         this.memoryContainerId = memoryContainerId;
-        this.memoryType = memoryType == null ? ShortTermMemoryType.CONVERSATION : memoryType;
         this.messages = messages;
-        this.binaryData = binaryData;
-        this.structuredData = structuredData;
+        this.data = data;
         this.namespace = namespace;
-        this.infer = infer; // default infer is false
         this.metadata = metadata;
-        this.tags = tags;
+
+        // Auto-detect memory type and set infer defaults
+        if (messages != null && !messages.isEmpty()) {
+            this.memoryType = ShortTermMemoryType.CONVERSATIONAL;
+            this.infer = infer; // User can override
+        } else if (data != null) {
+            this.memoryType = ShortTermMemoryType.DATA;
+            this.infer = false; // Always false for data
+        } else {
+            this.memoryType = memoryType; // Use provided type if neither messages nor data
+        }
+
         validate();
     }
 
     public void validate() {
+        // Check mutual exclusion
+        if (messages != null && !messages.isEmpty() && data != null) {
+            throw new IllegalArgumentException("Cannot specify both 'messages' and 'data' fields in the same request");
+        }
+
+        // Check at least one is provided
+        if ((messages == null || messages.isEmpty()) && data == null) {
+            throw new IllegalArgumentException("Must specify either 'messages' or 'data' field");
+        }
+
+        // Validate infer field for data type
+        if (data != null && infer) {
+            throw new IllegalArgumentException("infer=true is not supported for data memory type");
+        }
+
+        // Validate infer with messages
         if (messages == null || messages.isEmpty()) {
             if (infer) {
                 throw new IllegalArgumentException("No messages provided when inferring memory");
             }
         }
 
-        if (infer && memoryType != ShortTermMemoryType.CONVERSATION) {
-            throw new IllegalArgumentException("Infer is only supported for conversation memory");
+        if (infer && memoryType != ShortTermMemoryType.CONVERSATIONAL) {
```

Comment:
Seems too strict. We could do infer if the messages not null or empty. So if user want to preload some memory, they could use DATA type and set `messages` data.